unixfs/tar/writer.go

```diff
@@ -21,79 +21,100 @@ import (
 // TODO: does this need to be configurable?
 var DefaultBufSize = 1048576
 
-func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService, compression int) (io.Reader, error) {
+// DagArchive is equivalent to `ipfs getdag $hash | maybe_tar | maybe_gzip`
+func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService, useTar bool, compression int) (io.Reader, error) {
 
 	_, filename := path.Split(name)
 
 	// need to connect a writer to a reader
 	piper, pipew := io.Pipe()
 
-	// use a buffered writer to parallelize task
-	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
-
-	// construct the tar writer
-	w, err := NewWriter(bufw, dag, compression)
+	// compression determines whether to use gzip compression.
+	gzw, err := gzip.NewWriterLevel(pipew, compression)
 	if err != nil {
 		return nil, err
 	}
 
+	w := &Writer{
+		Dag: dag,
+		GzW: gzw,
+		ctx: ctx,
+	}
+
+	// construct the tar writer
+	if useTar {
+		w.TarW = tar.NewWriter(gzw)
+	}
+
 	// write all the nodes recursively
 	go func() {
-		if err := w.WriteNode(ctx, nd, filename); err != nil {
-			pipew.CloseWithError(err)
-			return
-		}
-
-		if err := bufw.Flush(); err != nil {
+		if err := w.WriteNode(nd, filename); err != nil {
 			pipew.CloseWithError(err)
 			return
 		}
 
+		w.Close()
 		pipew.Close() // everything seems to be ok.
 	}()
 
 	return piper, nil
 }
 
+func DagFS(ctx cxt.Context, nd *mdag.Node, dag mdag.DAGService, fpath string) error {
+	// Check if the output path already exists, so we know whether we should
+	// create our output with that name, or if we should put the output inside
+	// a preexisting directory
+	//rootExists := true
+	//rootIsDir := false
+	//if stat, err := os.Stat(fpath); err != nil && os.IsNotExist(err) {
+	//	rootExists = false
+	//} else if err != nil {
+	//	return err
+	//} else if stat.IsDir() {
+	//	rootIsDir = true
+	//}
+
+	w := &Writer{
+		Dag: dag,
+		ctx: ctx,
+	}
+	return w.WriteNode(nd, fpath)
+}
+
 // Writer is a utility structure that helps to write
 // unixfs merkledag nodes as a tar archive format.
 // It wraps any io.Writer.
 type Writer struct {
 	Dag  mdag.DAGService
 	TarW *tar.Writer
+	GzW  *gzip.Writer
+	ctx  cxt.Context
 }
 
-// NewWriter wraps given io.Writer.
-// compression determines whether to use gzip compression.
-func NewWriter(w io.Writer, dag mdag.DAGService, compression int) (*Writer, error) {
-
-	if compression != gzip.NoCompression {
-		var err error
-		w, err = gzip.NewWriterLevel(w, compression)
-		if err != nil {
-			return nil, err
+func (w *Writer) writeDir(nd *mdag.Node, fpath string) error {
+	if w.TarW != nil {
+		if err := writeDirHeader(w.TarW, fpath); err != nil {
+			return err
+		}
+	} else if w.GzW != nil {
+		// gz only case
+		return uio.ErrIsDir
+	} else {
+		// no compression / archive case
+		// thirdparty/tar/extractor.go uses 0755 but dir header in this file uses 0777
+		if err := os.MkdirAll(fpath, 0777); err != nil {
+			return err
```

Comment:
Sorry will put this back into initialization, just realized of a way to do this without using a "walkFn" with `interface{}`.